Recursive sample added for TreeNode iteratation

lavin-ds opened the pull request:
## Summary

TreeNode iteration documentation updated with an alternate non-recursive approach for traversing the tree using C#.

Fixes #1053

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/docs-desktop?pullRequest=1058) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docs-desktop?pullRequest=1058) before we can accept your contribution.<br/><br/>:x: lavin-ds [sign now](https://cla.dotnetfoundation.org/dotnet/docs-desktop?pullRequest=1058)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docs-desktop?pullRequest=1058) it.</sub>

lavin-ds commented:
I'm trying to ensure the ESLint rules pass the document but for some reason it catches multiple titles in the document, but I am unable to  find the issue.

adegeo reviewed (changes requested):
Thanks a bunch for contributing to the docs! There are a few things we need to handle though. Considering these articles were written maybe 10 years ago, our structure and code integrity checks have changed. By simply submitting the code projects with the PR, this repo will make sure the code can compile by running some actions on the PR.

If you need any help, or feel that it's too much to address the things I mentioned, let me know and I'll see if I can fit in some time to help out. Thanks again for contributing.

adegeo commented on dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md:
Adds the encoded `1 to represent the generic parameter and correctly link the xref

```suggestion
The following example is an alternate iterative approach to traversing the nodes of the tree using a <xref:System.Collections.Generic.Queue%601> data structure. This approach does not follow any ordering and only ensures all the nodes are printed. If we wish to follow a pre-order or post-order approach then we can use a <xref:System.Collections.Generic.Stack%601> data structure to ensure that a node and its children are processed before subsequent nodes.
```

lavin-ds replied:
Updated the document text to fix the xref as per your reference

lavin-ds commented on dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md:
```diff
@@ -17,6 +17,8 @@ It is sometimes useful to examine every node in a Windows Forms <xref:System.Win
  Each <xref:System.Windows.Forms.TreeNode> object in a tree view has properties that you can use to navigate the tree view: <xref:System.Windows.Forms.TreeNode.FirstNode%2A>, <xref:System.Windows.Forms.TreeNode.LastNode%2A>, <xref:System.Windows.Forms.TreeNode.NextNode%2A>, <xref:System.Windows.Forms.TreeNode.PrevNode%2A>, and <xref:System.Windows.Forms.TreeNode.Parent%2A>. The value of the <xref ...
   
+## Recursive approach
+
```
Removed extra heading

lavin-ds commented on dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md:
Added VB and CPP examples using the snippets

lavin-ds commented on dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md:
Indentation is now removed, code reference added via :::code::: reference blocks.

lavin-ds commented on dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md:
Added projects samples for all 3 languages in snippets folder and referenced in document via :::code::: blocks

adegeo commented:
@lavin-ds Things are looking great! It looks like we need to update our snippets testing system a bit to handle the C++ stuff. I'll get with my team and look into that. Besides that issue, things look great.

Three things I noticed that can be changed in the article to make it better:

1. The first section, recursive, uses a numbered step approach in introducing the code. The non-recursive section doesn't. I like your approach better for the non-recursive. Can you remove the steps and just introduce the code in the recursive section? Then both sections match in style.
2. The article repeats this clause `recursive procedure (recursive method in C# and C++)` but really it should be reversed. It should say `recursive method (recursive procedure in VB.NET)`. I also don't think we need it mentioned twice. Simply stating `recursive method` the 2nd time would be adequate.

Thanks again for your contribution. Let me ask you, how easy was it to figure out the `:::Code` and snippets folder format? I'm always hesitant to ask more of our public contributors when it comes to having to migrate code to the new system.

lavin-ds commented:
@adegeo Great to hear that 👍🏽 

1. & 2. I have updated the document to match the style of both the sections and also changed the method/procedure to remove redundancy.
3. I actually originally wanted to go with a .NET5 using vsCode approach, but since the parent folder of this control was called framework I decided to go with VS and .Net framework. A new framework solution in VS automatically follows the old project style.

Overall, understanding the :::Code format and snippet structure was easy once I followed the existing samples provided by you. The challenging part for me was creating the C++ solution since I had no prior experience with it 👯

adegeo commented:
@lavin

lavin-ds commented:
@adegeo  are there any further changes required on  the PR?

adegeo commented:
@lavin-ds It looks like the CPP project was tested correctly, but I think we need to force visual studio to run the project (instead of dotnet)

Can you add a `snippets.5000.json` file in the `cpp` directory (The directory with the project.vcxproj file) with this content?

```json
{
    "host": "visualstudio"
}
```

Commit that and then that will force the snippets to run again.

lavin-ds commented:
I've added the snippet file with the host as visual studio, this should help with targeting the build

adegeo reviewed (approved):
Thanks for helping us figure out CPP processing 😁 Looks good!

lavin-ds commented:
Thanks for your help with this too, it was a lot of fun working on this and contributing to the project. I should be comfortable picking up more of these now!! 

How does the merge happen? Or will you do it based on a sprint or release plan?

Thanks again for your help with figuring this out 👍🏽

adegeo commented on dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md:
Moved the warning into the recursive section. Cleaned this up a bit. I think it's a bit more clear now, thoughts?

```suggestion
The following example is an alternate iterative approach to traversing the nodes of the tree using a <xref:System.Collections.Generic.Queue%601> collection. This approach doesn't follow the parent-child relationship of a node only ensures every node is printed. If you want to process each tree node and its children, first, use the <xref:System.Collections.Generic.Stack%601> collection
```
